xdsl/dialects/stencil.py

```diff
@@ -458,16 +463,27 @@ def print_assign_argument(args: tuple[BlockArgument, SSAValue, Attribute]):
             printer.print(" : ")
             printer.print(args[2])
 
+        def print_destination_operand(dest: SSAValue):
+            printer.print(dest)
+            printer.print(" : ")
+            printer.print(dest.type)
+
         printer.print("(")
         printer.print_list(
             zip(self.region.block.args, self.args, (a.type for a in self.args)),
             print_assign_argument,
         )
         printer.print(") -> (")
-        printer.print_list((r.type for r in self.res), printer.print_attribute)
+        if self.dest:
+            printer.print_list(self.dest, print_destination_operand)
+        else:
+            printer.print_list((r.type for r in self.res), printer.print_attribute)
```

Comment:
Not so keen on this syntax, if I'm understanding correctly that it hides that the bufferized version does not actually have a SSA return value.

Also, can you not have dest+res in the unbufferized mode?

Reply:
I went with apply only able to be fully value-semantic (having results in res) or reference-semantics (having destinations in dest), never a mix of both

Reply:
I could clarify that in a docstring I guess

Reply:
My point was something else, that dest is printed after the `->` (which unhelpfully isn't shown in the 4-line snippet above but comes just before)

Reply:
To keep it simple, you could have `-> ( ... )` for results and `dest( .... )` or `outs(....)` for the destination-style passing.

Reply:
You could also consider re-using `print_assign_argument`

Reply:
Done with outs, WDYT?

Reply:
lgtm, thanks for this